src/experimental/tools/download2go/apis/dash/dashSegmentsBuilder.ts

```diff
@@ -0,0 +1,533 @@
+              utils: {
+                contentID,
+                id,
+                segmentKey: [`${id}--${contentID}--${i}`, i],
+                type,
+                url: mediaURL || "",
+              },
+            });
+          }
+          return {
+            init: key,
+            segments: segmentsBuilder,
+          };
+        })
+      );
+    })
+  );
+};
+/**
+ * Get the init segments buffer from a server
+ *
+ * @param ISegment - index segment we use to build segments
+ * @returns An Observable of init segment get on the server
+ *
+ */
+const getInitSegments = (Segment: ISegment): Observable<ISegmentIndex> => {
+  const { indexRange, range, mediaURL, duration = 0 } = Segment;
+  if (indexRange && range) {
+    return forkJoin({
+      initSegment: xhrRequest({
+        url: mediaURL || "",
+        headers: { Range: `bytes=${range.join("-")}` },
+        responseType: "arraybuffer",
+      }).pipe(mergeMap(({ value }) => of(value.responseData))),
+      indexSegment: xhrRequest({
+        url: mediaURL || "",
+        headers: { Range: `bytes=${indexRange.join("-")}` },
+        responseType: "arraybuffer",
+      }).pipe(mergeMap(({ value }) => of(value.responseData))),
+    }).pipe(
+      mergeMap(({ initSegment, indexSegment }) => {
+        const dataInit = concatBytes(
+          new Uint8Array(initSegment),
+          new Uint8Array(indexSegment)
+        );
+        const nextSegmentsRanges = getSegmentsFromSidx(
+          dataInit,
+          indexRange ? indexRange[0] : 0
+        );
+        return of({
+          dataInit,
+          duration,
+          mediaURL,
+          nextSegmentsRanges,
+          type: "BaseRepresentationIndex" as const,
+        });
+      })
+    );
+  }
+  return xhrRequest({
+    url: mediaURL || "",
+    responseType: "arraybuffer",
+  }).pipe(
+    mergeMap(({ value }) =>
+      of({
+        dataInit: value.responseData,
+        mediaURL,
+        type: "TemplateRepresentationIndex" as const,
+      })
+    )
+  );
+};
+
+/**
+ * A download manager that take all the representations to execute them
+ * concurrently and tranform a Segment non built into a segment built
+ *
+ * @param utilsBuilder - Emitter and db instances
+ * @param progress$ - A Subject observable to emit related with the progress percentage
+ * @param pause$ - A Subject observable to emit when the user pause the download
+ * @param contentID - A string that define the content id of the current download
+ * @returns An Observable of rxpManifest flow
+ *
+ */
+export const downloadManager = (
+  utilsBuilder: IUtils,
+  progress$: ReplaySubject<IProgressBarBuilderAbstract>,
+  pause$: AsyncSubject<void>,
+  contentID: string
+) => (
+  source$: Observable<ILocalManifestOnline>
+): Observable<IDownloadManagerOutput> =>
+  source$.pipe(
+    concatMap(manifest => {
+      if (!manifest) {
+        return from([]);
+      }
+      return from(manifest.periods).pipe(
```

Comment:
I'm a little bit uncomfortable with the fact of creating an observable out of an object just for the sake of creating an observable flow. You could just make a loop on periods / adaptations / representations / segments to get the createSegment observable, aggregate the observables in order to make a unique obs.

Reply:
The problem is:
- I use this flow to store segments when they arrive and then at some key moment, I store the whole manifest to record the global progress,  but IndexDB doesnt handle complex object but just basic type thus I can't store observable type in IndexDB.